### PR TITLE
create ui and show user location on map

### PR DESCRIPTION
**Pull request by ghost**
## Related Tickets
- [#Task 75362: Map screen](https://edu-redmine.sun-asterisk.vn/issues/75362)
## WHAT
- Create UI Mapscreen
- Display map
- Show user location on map
## Evidence (Screenshot or Video)
Screenshot 2024-05-07 at 11 15 30

## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://sal.vn/Zts4P8 |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Message | Sun* commit message | Template: [TASK/FREATURE/BUG/SUPPORT][RedmineID] + message |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://sal.vn/piPoQn |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: Weather_App/Weather_App/Scenes/Map/LocationManager.swift [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    private let locationManager = CLLocationManager()
+    private var currentLocationSubject = PublishSubject<CLLocation>()
+    private var authorizationStatusSubject = BehaviorSubject<CLAuthorizationStatus>(value: .notDetermined)
+    
+    private override init() {
+        super.init()
+        locationManager.delegate = self
+        locationManager.requestWhenInUseAuthorization()
+        locationManager.desiredAccuracy = kCLLocationAccuracyBest
+        locationManager.distanceFilter = kCLDistanceFilterNone
```
Comment (dongtg-1485) [inline Weather_App/Weather_App/Scenes/Map/LocationManager.swift]:
```suggestion
        locationManager.do {
            // Code
        }
```

**Review by dongtg-1485 — APPROVED** [review]

File: Weather_App/Weather_App/Scenes/Map/MapViewController.swift [hunk 2]
```diff
@@ -7,19 +7,53 @@
...
 
 final class MapViewController: UIViewController {
-
+    
+    @IBOutlet weak var mapView: MKMapView!
+    @IBOutlet weak var weatherView: UIView!
+    @IBOutlet weak var statusView: UIView!
+    @IBOutlet weak var imageStatus: UIImageView!
+    @IBOutlet weak var nameCityLabel: UILabel!
+    @IBOutlet weak var temperatureLabel: UILabel!
```
Comment (ghost) [inline Weather_App/Weather_App/Scenes/Map/MapViewController.swift]: private e

File: Weather_App/Weather_App/Scenes/Map/MapViewController.swift [hunk 3]
```diff
@@ -7,19 +7,53 @@
...
 import RxSwift
+import MapKit
 
 final class MapViewController: UIViewController {
-
+    
+    @IBOutlet weak var mapView: MKMapView!
+    @IBOutlet weak var weatherView: UIView!
+    @IBOutlet weak var statusView: UIView!
+    @IBOutlet weak var imageStatus: UIImageView!
```
Comment (ghost) [inline Weather_App/Weather_App/Scenes/Map/MapViewController.swift]: statusImageView

File: Weather_App/Weather_App/Scenes/Map/MapViewController.swift [hunk 4]
```diff
@@ -7,19 +7,53 @@
...
+        navigationItem.searchController = searchController
+        customizeNavigationController()
+        customizeSearchController(searchController: searchController)
+        definesPresentationContext = true
     }
 }
 
 extension MapViewController: Bindable {
     func bindViewModel() {
+        let input = MapViewModel.Input(getCurrentLocationTrigger: Observable.just(()).asDriver(onErrorJustReturn: ()))
```
Comment (ghost) [inline Weather_App/Weather_App/Scenes/Map/MapViewController.swift]: dùng `Driver.just(())` cho gọn e

File: Weather_App/Weather_App/Scenes/Map/MapViewController.swift [hunk 5]
```diff
@@ -7,19 +7,53 @@
...
+        let output = viewModel.transform(input: input, disposeBag: disposeBag)
+        
+        output.currentLocation
+            .drive(onNext: { [weak self] location in
+                self?.locationManager.setRegion(on: self?.mapView ?? MKMapView(),
+                                                center: location.coordinate,
+                                                latitudinalMeters: Constants.latitudinalMeters,
+                                                longitudinalMeters: Constants.longitudinalMeters)
+            })
```
Comment (ghost) [inline Weather_App/Weather_App/Scenes/Map/MapViewController.swift]: tách logic trong drive này thành binder cho gọn e

File: Weather_App/Weather_App/Scenes/Map/MapViewModel.swift [hunk 6]
```diff
@@ -18,12 +19,20 @@ struct MapViewModel {
...
     }
     
     struct Output {
+        let currentLocation: Driver<CLLocation>
     }
 
     func transform(input: Input, disposeBag: DisposeBag) -> Output {
-        return Output()
+        let currentLocation = input.getCurrentLocationTrigger
+            .flatMapLatest { _ in
```
Comment (ghost) [inline Weather_App/Weather_App/Scenes/Map/MapViewModel.swift]:
```suggestion
            .flatMapLatest {
```

File: Weather_App/Weather_App/Scenes/Map/MapViewController.swift [hunk 7]
```diff
@@ -7,19 +7,63 @@
...
+            .disposed(by: disposeBag)
     }
 }
 
+extension MapViewController {
+    private func bindLocationUpdate() -> Binder<CLLocation> {
+        return Binder(self) { viewController, location in
+            viewController.updateMapRegion(location)
+        }
+    }
```
Comment (ghost) [inline Weather_App/Weather_App/Scenes/Map/MapViewController.swift]: dùng property là được rồi e, k cần func, ở line 52 sửa thành `.drive(to: locationUpdateBinder)`
```suggestion
    private var locationUpdateBinder: Binder<CLLocation> {
        return Binder(self) { vc, location in
            vc.updateMapRegion(location)
        }
    }
```

File: Weather_App/Weather_App/Scenes/Map/MapViewController.swift [hunk 8]
```diff
@@ -7,19 +7,63 @@
...
+    private func updateMapRegion(_ location: CLLocation) {
+        locationManager.setRegion(on: mapView,
+                                   center: location.coordinate,
+                                   latitudinalMeters: Constants.latitudinalMeters,
+                                   longitudinalMeters: Constants.longitudinalMeters)
+    }
```
Comment (ghost) [inline Weather_App/Weather_App/Scenes/Map/MapViewController.swift]: tạo thêm 1 cái ext cho thằng này, k nên để chung với Binder

File: Weather_App/Weather_App/Utils/CustomeNavigationController.swift [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+                searchIcon.image = searchIcon.image?.withRenderingMode(.alwaysTemplate)
+                searchIcon.tintColor = .black
+            }
+        }
+        searchController.searchBar.backgroundColor = .blueGreen
+        searchController.searchBar.addShadow()
+        UIBarButtonItem.appearance(whenContainedInInstancesOf: [UISearchBar.self]).tintColor = .white
+    }
+}
+
```
Comment (ghost) [inline Weather_App/Weather_App/Utils/CustomeNavigationController.swift]: remove blank

**Review by ghost — APPROVED** [review]

**Review by daolq-2712 — APPROVED** [review]
OK!

**Pull request merged**